rfcs/55-deterministic_rng.md

```diff
@@ -0,0 +1,135 @@
+
+
+## Drawbacks
+
+- This may not be general enough to include in Bevy.
+- This may not be general enough to be on by default.
+- Includes `rand` as a dependency
+  - But not in public API.
+
+## Rationale and alternatives
+
+- Why is this design the best in the space of possible designs?
+  - It gives flexibility for PRNG crates while also enforcing standards on the ecosystem. 
+  - It defaults to what someone would resonably expect if they wrote it themselves.
+  - It defaults to something safe (suitable for cryptographic functions), removing a possible footgun.
+- What other designs have been considered and what is the rationale for not choosing them?
+  - We could go higher and expose an API closer to `rand`. This is what [Godot](https://docs.godotengine.org/en/stable/tutorials/math/random_number_generation.html), [Unity](https://docs.unity3d.com/ScriptReference/Random.html), and [Unreal](https://docs.unrealengine.com/4.27/en-US/BlueprintAPI/Math/Random/) do. We are not experts in PRNG API design, and while `rand` is clearly the most popular random crate in the Rust ecosystem we don't currently want to tie bevy to any particular API in case something better emerges.
+  - We could go lower and merely expose a static `WorldSeed`. I'm worried about what the default would be and forcing a seed at world creation feels heavyweight.
+  - We could default to a faster PRNG rather than a safer one. I wanted folks to fall into the pit of success.
+- What objections immediately spring to mind? How have you addressed them?
+  - `rand` is too heavy of a dependency.
+  - This should not be in core.
+- What is the impact of not doing this?
+  - There is no chance of the Bevy ecosystem supporting deterministic execution.
+- Why is this important to implement as a feature of Bevy itself, rather than an ecosystem crate?
+  - There needs to be one true way to keep the ecosystem coherant.
+
+## Prior art
+
+* https://github.com/bevyengine/bevy/discussions/2480
+* https://github.com/bevyengine/bevy/discussions/1678
+
+Other engines such as [Godot](https://docs.godotengine.org/en/stable/tutorials/math/random_number_generation.html), [Unity](https://docs.unity3d.com/ScriptReference/Random.html), and [Unreal](https://docs.unrealengine.com/4.27/en-US/BlueprintAPI/Math/Random/) include engine support for randomness (with a higher-level API).
+
+* https://towardsdatascience.com/how-to-use-random-seeds-effectively-54a4cd855a79
+* https://www.whatgamesare.com/determinism.html
+* https://gafferongames.com/post/deterministic_lockstep/
+* https://ruoyusun.com/2019/03/29/game-networking-2.html
+* https://arxiv.org/abs/2104.06262
+
+## Unresolved questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+  - Do we want this?
+  - What about `rand`?
+  - Is `Entropy` naming too obscure?
+  - How do we deal with security vs speed tradeoff here?
```

Comment:
This question here is really important, and given how "Too Much Security" can have an impact on performance/throughput, we probably need to ask which is the most common case in needing a source of randomness. There *will* be cases needing more cryptographically secure sources. But for randomising damage rolls, etc, there won't really be a need for something super secure, just random enough and fast. If more use-cases favour needing fast sources than secure sources, the default should be biased towards fast, and secure sources could be an optional feature to be enabled when needed. The random/entropy crate should be well documented and put this on the fore-front about such differences/needs.

Reply:
This is merely the initial entropy pool. So for apps that don't need secure randomness, there is one initial slower call and that can be used to seed a faster rng. Then again, it could be an enormous perf footgun if someone does the obvious thing and just uses one pool. I'm not sure if I want a more common perf footgun or a less common security footgun 🤔.

Reply:
I basically created `bevy_turborand` to explore this, by exposing various Rng structs while making them threadsafe. There's a `GlobalRng` as a resource to be used for the explicit purpose of seeding `RngComponent`s. As long as `GlobalRng` is given a specific seed, then everything becomes deterministic from the source.

For non-deterministic Rng, I use instead thread_local Rngs initialised with a generate entropy function, and seed new Rng instances from that. Basically same mechanism as with `fastrand` does.

The end result of this should be Rngs that are seeded very quickly and with deterministically derived random states. At the moment, it's just `WyRand` so nothing that is cryptographically secure. But I hope to add ChaCha8 eventually and maybe play around with buffering so to increase throughput as well.

Reply:
Awesome! That is similar to my vision of where this should go in future rfcs. The fact that most engines in the industry provide rng's means bevy should too, and this building block is a lower level layer that folks can drop down to if they need to customize anything. But it keeps the ecosystem potentially deterministic if both the baked in rngs and any plugin rngs decend from the same seed

Reply:
But I am not sure a thread local RNG is the right grouping fwiw...I like the idea of tying to systems more.

Reply:
`rand` and `fastrand` both have mechanisms to create an rng instance from a thread-local seed. This mechanism I would describe as _non-deterministic_ because the intent is to provide as random state (and since these rngs are not threadsafe, they are provided in a thread local state). Which is fine for a lot of cases.

Now, one might want that globally, if given no seed, you derive a random seed (which can be generated in the above manner), and this would be how the `GlobalRng` gets initialised. From there, every other `RngComponent` gets its own seeded state deterministically from `GlobalRng`. So only one point needed to insert randomness while everything else is then deterministic.